middle_end/flambda/simplify/simplify_simple.ml

```diff
@@ -20,87 +20,47 @@ module DA = Downwards_acc
-    | Bottom -> Bottom, T.bottom (T.kind ty)
-    | Ok ty -> Ok simple, ty
-
-let simplify_simple dacc simple ~min_name_mode : _ Or_bottom.t * _ =
+let simplify_simple dacc simple ~min_name_mode =
   let typing_env = DA.typing_env dacc in
-  match
-    TE.get_canonical_simple_with_kind_exn typing_env simple ~min_name_mode
-  with
+  match TE.type_simple_in_term_exn typing_env simple ~min_name_mode with
   | exception Not_found ->
     Misc.fatal_errorf "No canonical [Simple] for %a exists at the@ \
         requested name mode (%a) or one greater.@ Downwards accumulator:@ %a"
       Simple.print simple
       Name_mode.print min_name_mode
       DA.print dacc
-  | simple, kind -> type_for_simple' simple kind
-
-type changed =
-  | Unchanged
-  | Changed
+  | ty -> ty
 
-let simplify_simples dacc simples ~min_name_mode =
-  let typing_env = DA.typing_env dacc in
-  let changed = ref Unchanged in
-  let result =
-    Or_bottom.all (List.map (fun simple : _ Or_bottom.t ->
-        match
-          TE.get_canonical_simple_with_kind_exn typing_env simple
-            ~min_name_mode
-        with
-        | new_simple, kind ->
-          if new_simple != simple then begin
-            changed := Changed;
-          end;
-          type_for_simple new_simple kind
-        | exception Not_found ->
-          Misc.fatal_errorf "No canonical [Simple] for %a exists at the@ \
-              requested name mode (%a) or one greater.@ \
-              Downwards accumulator:@ %a"
-            Simple.print simple
-            Name_mode.print min_name_mode
-            DA.print dacc)
-      simples)
-  in
-  !changed, result
+type simplify_simples_result = {
+  seen_bottom : bool;
+  simples : Simple.t list;
+  simple_tys : Flambda_type.t list;
+}
 
-let simplify_simples' dacc simples ~min_name_mode =
+let simplify_simples dacc simples =
   let typing_env = DA.typing_env dacc in
-  let changed = ref Unchanged in
-  let result =
-    Or_bottom.all (List.map (fun simple : _ Or_bottom.t ->
-        match TE.get_canonical_simple_exn typing_env simple ~min_name_mode with
-        | new_simple ->
-          if new_simple != simple then begin
-            changed := Changed;
-          end;
-          Ok new_simple
-        | exception Not_found ->
-          Misc.fatal_errorf "No canonical [Simple] for %a exists at the@ \
-              requested name mode (%a) or one greater.@ \
-              Downwards accumulator:@ %a"
-            Simple.print simple
-            Name_mode.print min_name_mode
-            DA.print dacc)
-      simples)
+  let seen_bottom = ref false in
+  let simple_tys =
+    ListLabels.map simples ~f:(fun simple ->
+      match
+        TE.type_simple_in_term_exn typing_env simple
+          ~min_name_mode:Name_mode.normal
+      with
+      | ty ->
+        (* [ty] will always be an alias type; see the implementation of
+           [TE.get_canonical_simple_in_term_exn]. *)
+        if T.is_obviously_bottom ty then begin
```

Comment:
See comment above: an alias is never considered obviously bottom, so this branch is dead code.

Reply:
Should be fixed in the latest commit.